# Wire latest BakaSU preinit and module blacklist into the prepared 4.9 SUSFS kernel.
# Apply after susfs-inline-hook.sh. Preserve pre-load permission checks and
# retry the original init after a recoverable preinit exec failure.

diff --git a/fs/exec.c b/fs/exec.c
index e664837..10bacb0 100644
--- a/fs/exec.c
+++ b/fs/exec.c
@@ -1689,6 +1689,11 @@ extern int ksu_handle_execveat_sucompat(int *fd, struct filename **filename_ptr,
 				 void *argv, void *envp, int *flags);
 extern int ksu_handle_post_execveat_sucompat(int *fd, struct filename **filename_ptr,
 			void *argv, void *envp, int *flags, int *retval);
+extern int ksu_setup_preinit_file(struct filename **filename_ptr,
+			struct user_arg_ptr *argv);
+extern int ksu_preinit_handle_post_execveat(int *fd, struct filename **filename_ptr,
+			void *argv, void *envp, int *flags, int *retval,
+			int *preinit_fd);
 #endif
 
 static int do_execveat_common(int fd, struct filename *filename,
@@ -1704,10 +1709,17 @@ static int do_execveat_common(int fd, struct filename *filename,
 
 #ifdef CONFIG_KSU_SUSFS
 	bool is_su_session = false;
+	int preinit_fd = -1;
+
+ksu_exec_retry:
+	pathbuf = NULL;
+	is_su_session = false;
 #endif
 	if (IS_ERR(filename))
 		return PTR_ERR(filename);
 #ifdef CONFIG_KSU_SUSFS
+	if (unlikely(current->pid == 1))
+		preinit_fd = ksu_setup_preinit_file(&filename, &argv);
 	if (likely(susfs_is_current_proc_no_su()))
 		goto orig_flow;
 	if (static_branch_likely(&ksu_su_compat_enabled)) {
@@ -1836,6 +1848,9 @@ orig_flow:
 #ifdef CONFIG_KSU_SUSFS
 	if (unlikely(is_su_session))
 		(void)ksu_handle_post_execveat_sucompat(&fd, &filename, &argv, &envp, &flags, &retval);
+	if (unlikely(preinit_fd >= 0))
+		(void)ksu_preinit_handle_post_execveat(&fd, &filename, &argv,
+				&envp, &flags, &retval, &preinit_fd);
 #endif
 	current->fs->in_exec = 0;
 	current->in_execve = 0;
@@ -1866,6 +1881,13 @@ out_files:
 	if (displaced)
 		reset_files_struct(displaced);
 out_ret:
+#ifdef CONFIG_KSU_SUSFS
+	if (unlikely(preinit_fd >= 0) &&
+	    ksu_preinit_handle_post_execveat(&fd, &filename, &argv, &envp,
+			&flags, &retval, &preinit_fd) < 0 &&
+	    !fatal_signal_pending(current))
+		goto ksu_exec_retry;
+#endif
 	putname(filename);
 	return retval;
 }
diff --git a/kernel/module.c b/kernel/module.c
index f42fbf1..a5b6488 100644
--- a/kernel/module.c
+++ b/kernel/module.c
@@ -3825,6 +3825,11 @@ static int load_module(struct load_info *info, const char __user *uargs,
 	return err;
 }
 
+#ifdef CONFIG_KSU_SUSFS
+extern void ksu_handle_init_module(int *result);
+extern void ksu_handle_finit_module(int *result);
+#endif
+
 SYSCALL_DEFINE3(init_module, void __user *, umod,
 		unsigned long, len, const char __user *, uargs)
 {
@@ -3842,7 +3847,11 @@ SYSCALL_DEFINE3(init_module, void __user *, umod,
 	if (err)
 		return err;
 
-	return load_module(&info, uargs, 0);
+	err = load_module(&info, uargs, 0);
+#ifdef CONFIG_KSU_SUSFS
+	ksu_handle_init_module(&err);
+#endif
+	return err;
 }
 
 SYSCALL_DEFINE3(finit_module, int, fd, const char __user *, uargs, int, flags)
@@ -3869,7 +3878,11 @@ SYSCALL_DEFINE3(finit_module, int, fd, const char __user *, uargs, int, flags)
 	info.hdr = hdr;
 	info.len = size;
 
-	return load_module(&info, uargs, flags);
+	err = load_module(&info, uargs, flags);
+#ifdef CONFIG_KSU_SUSFS
+	ksu_handle_finit_module(&err);
+#endif
+	return err;
 }
 
 static inline int within(unsigned long addr, void *start, unsigned long size)
